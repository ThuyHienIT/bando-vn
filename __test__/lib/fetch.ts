const mockFetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({}),
    text: () => Promise.resolve(''),
    clone: () => ({
      json: () => Promise.resolve({}),
      text: () => Promise.resolve(''),
    }),
  })
);

if (!global.fetch) {
  (global as any).fetch = mockFetch;
}

export { mockFetch };
