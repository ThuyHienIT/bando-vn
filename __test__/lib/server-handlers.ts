// this is put into here so I can share these same handlers between my tests

const handlers = [] as any;

export { handlers };
