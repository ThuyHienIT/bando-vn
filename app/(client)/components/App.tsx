'use client';

import { ConfigProvider } from 'antd';
import { memo } from 'react';
import { ThemeProvider } from 'styled-components';

import { configProviders } from '../antd-theme';
import { theme } from '../theme';

interface Props {
  children: React.ReactNode;
}
export const App = memo<Props>(function BasicLayout(props) {
  return (
    <ThemeProvider theme={theme}>
      <ConfigProvider {...configProviders}>{props.children}</ConfigProvider>
    </ThemeProvider>
  );
});
