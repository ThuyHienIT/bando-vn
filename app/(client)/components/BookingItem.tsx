'use client';

import { Button, Card, Typography } from 'antd';
import dayjs from 'dayjs';
import Image from 'next/image';
import { useCallback } from 'react';

import { EditOutlined } from '@ant-design/icons';

interface Props {
  data: BookingItem;
  onCancel?(data?: BookingItem): void;
  onEdit?(data?: BookingItem): void;
}

export function BookingItem({ onCancel, onEdit, ...props }: Props) {
  const handleCancel = useCallback(() => {
    onCancel?.(props.data);
  }, [onCancel, props.data]);

  const handleEdit = useCallback(() => {
    onEdit?.(props.data);
  }, [onEdit, props.data]);

  return (
    <>
      <Card
        size="small"
        cover={
          props.data.facility?.thumbnail ? (
            <Image
              alt={props.data.facility.name}
              src={props.data.facility?.thumbnail}
              width={100}
              height={100}
            />
          ) : null
        }
        actions={[
          <Button
            key="details"
            type="text"
            size="small"
            danger
            onClick={handleCancel}
          >
            Cancel
          </Button>,
          <Button
            icon={<EditOutlined />}
            type="text"
            size="small"
            key="edit"
            onClick={handleEdit}
          >
            Edit
          </Button>,
        ]}
      >
        <Typography.Title level={5}>
          {props.data.facility?.name}
        </Typography.Title>
        <Typography.Text ellipsis>{`From: ${dayjs(props.data.from).format(
          'DD/MM/YYYY HH:mm'
        )}`}</Typography.Text>
        <Typography.Text ellipsis>{`To: ${dayjs(props.data.to).format(
          'DD/MM/YYYY HH:mm'
        )}`}</Typography.Text>
      </Card>
    </>
  );
}
