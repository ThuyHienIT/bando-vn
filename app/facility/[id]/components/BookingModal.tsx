'use client';

import {
  Button,
  DatePicker,
  Form,
  Modal,
  notification,
  TimePicker
} from 'antd';
import { useForm } from 'antd/lib/form/Form';
import request from 'app/(client)/lib/request';
import { Dayjs } from 'dayjs';
import { useCallback, useEffect, useState } from 'react';
import styled from 'styled-components';

import { ExclamationCircleFilled } from '@ant-design/icons';

const Inputs = styled(Form.Item)`
  .ant-form-item-control-input-content {
    display: flex;
    gap: 8px;
  }
`;

interface Props {
  className?: string;
  from?: Dayjs;
  to?: Dayjs;
  data?: FacilityItem;
  opened: boolean;
  isUpdate?: boolean;
  bookingId?: string;
  onClose(): void;
  onDone?(item?: BookingItem): void;
  onCancelled?(bookingId?: string): void;
}

type FormValues = {
  date: Dayjs;
  from: Dayjs;
  to: Dayjs;
};
export function BookingModal({
  onClose,
  onDone,
  onCancelled,
  ...props
}: Props) {
  const [form] = useForm<FormValues>();
  const [loading, setLoading] = useState(false);
  const [api, contextHolder] = notification.useNotification();

  const onSubmit = useCallback(
    async (values: FormValues) => {
      try {
        setLoading(true);

        const date = values.date
          .clone()
          .set('hour', 0)
          .set('minute', 0)
          .set('second', 0);

        const url = props.isUpdate
          ? '/api/booking/update'
          : '/api/facility/reserve';

        const resp = await request(url, {
          method: 'POST',
          body: JSON.stringify({
            facilityId: props.data?.id,
            from: date
              .clone()
              .add(values.from.get('hour'), 'hour')
              .add(values.from.get('minute'), 'minute'),
            to: date
              .clone()
              .add(values.to.get('hour'), 'hour')
              .add(values.to.get('minute'), 'minute'),
            userEmail: 'kqthang1505@gmail.com',
            ...(props.bookingId ? { id: props.bookingId } : {}),
          }),
        });

        api.success({
          message: props.isUpdate
            ? 'Update booking successful'
            : `Reserve a slot on ${values.date.format(
                'DD MMM YYYY'
              )}, from ${values.from.format('HH:mm')} - to ${values.to.format(
                'HH:mm'
              )} successfully`,
        });
        onDone?.(resp);
      } catch (e: any) {
        console.log('e', e);
        api.error({ message: e.message });
      } finally {
        setLoading(false);
      }
    },
    [props.isUpdate, props.data?.id, props.bookingId, api, onDone]
  );

  const handleOK = useCallback(() => {
    form.submit();
  }, [form]);

  const cancelBooking = useCallback(() => {
    Modal.confirm({
      title: `Cancel your booking`,
      icon: <ExclamationCircleFilled />,
      content: `Are you sure you want to cancel your booking of ${props.data?.name}?`,
      onOk: async () => {
        try {
          await request(`/api/booking/cancel/${props.bookingId}`);
          api.success({ message: 'Cancel booking successful' });

          onCancelled?.(props.bookingId);
        } catch (e: any) {}
      },
      onCancel() {},
      okText: 'Yes',
      cancelText: 'No',
    });
  }, [api, onCancelled, props.bookingId, props.data?.name]);

  const handleCancel = useCallback(() => {
    if (!props.isUpdate) {
      onClose();
      return;
    }

    if (props.data) cancelBooking();
  }, [cancelBooking, onClose, props.data, props.isUpdate]);

  useEffect(() => {
    if (props.opened)
      form.setFieldsValue({ from: props.from, to: props.to, date: props.from });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.opened]);

  return (
    <>
      {contextHolder}
      <Modal
        rootClassName="booking-modal"
        open={props.opened}
        onCancel={onClose}
        title={`Booking this slot of ${props.data?.name}`}
        destroyOnClose
        footer={
          <>
            <Button
              data-testid="booking-modal-cancel-btn"
              className="confirm-modal-no-btn"
              onClick={props.isUpdate ? handleCancel : onClose}
            >
              {props.isUpdate ? 'Cancel this booking' : 'Cancel'}
            </Button>
            <Button
              data-testid="booking-modal-yes-btn"
              type="primary"
              onClick={handleOK}
              loading={loading}
            >
              {props.isUpdate ? 'Update' : 'Book'}
            </Button>
          </>
        }
      >
        <Form
          data-testid="booking-form"
          form={form}
          initialValues={{
            from: props.from,
            to: props.to,
            date: props.from,
          }}
          layout="vertical"
          onFinish={onSubmit}
        >
          <Inputs label="Date" style={{ gap: 8 }}>
            <Form.Item name="date" noStyle rules={[{ required: true }]}>
              <DatePicker showToday={false} />
            </Form.Item>
            <Form.Item name="from" noStyle rules={[{ required: true }]}>
              <TimePicker showSecond={false} format="HH:mm" showNow={false} />
            </Form.Item>
            <Form.Item name="to" noStyle rules={[{ required: true }]}>
              <TimePicker
                showSecond={false}
                format="HH:mm"
                showNow={false}
                use12Hours={false}
              />
            </Form.Item>
          </Inputs>
        </Form>
      </Modal>
    </>
  );
}
