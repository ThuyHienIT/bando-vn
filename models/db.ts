import { randomUUID } from 'crypto';
import dayjs from 'dayjs';
import postgres from 'postgres';

const sql = postgres({ ssl: true });

type BaseType = {};
type DataType = BaseType & Record<string, string | number | boolean>;

export async function insertOne<T extends BaseType>(
  table: string,
  data: DataType
) {
  data.id = randomUUID();
  data.created_at = dayjs().format();
  data.updated_at = dayjs().format();

  const resp = await sql`
    INSERT INTO ${sql(`${table}`)} ${sql(data)}
    returning *
  `;

  return resp[0] as T;
}

export async function insertMany<T extends BaseType>(
  table: string,
  datas: DataType[]
) {
  datas.forEach((data) => {
    data.id = randomUUID();
    data.created_at = dayjs().format();
    data.updated_at = dayjs().format();
  });

  const resp: unknown = await sql`
    INSERT INTO ${sql(`${table}`)} ${sql(datas)}
    returning *
  `;

  return resp as T[];
}

export async function queryAll<T extends BaseType>(
  table: string,
  fields: string[] = []
) {
  const resp: unknown = await sql`
    SELECT ${fields.length > 0 ? sql(fields) : sql`*`} FROM ${sql(`${table}`)}
  `;

  return resp as T[];
}

export async function queryById<T extends BaseType>(table: string, id: string) {
  const data = await sql`
    SELECT * FROM ${sql(table)} WHERE id=${id}
  `;
  return data[0] as T;
}

export async function deleteByIds(table: string, ids: string[]) {
  await sql`
    DELETE FROM ${sql(table)} WHERE id in ${sql(ids)}
  `;
  return true;
}

export async function updateOne<T extends BaseType>(
  table: string,
  data: DataType
) {
  const { id, updated_at, ...d } = data;

  const updated = await sql`
    UPDATE ${sql(table)} SET ${sql(d)},
    updated_at=now()
    WHERE id=${id}`;

  return updated[0] as T;
}

export { sql };
