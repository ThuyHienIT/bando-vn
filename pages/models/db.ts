import path from 'path';
import { promises as fs } from 'fs';
import dayjs from 'dayjs';
import { randomUUID } from 'crypto';
import { tryParseJson } from '../lib/tryParseJSON';

const DB_FOLDER = process.env.NODE_ENV === 'test' ? 'dbtest' : 'db';
const DB_DIR = path.join(process.cwd(), DB_FOLDER);

type RecordType = { id: string; createdAt?: string; updatedAt?: string };

async function prepareDb(dbName: string) {
  const filePath = path.resolve(DB_DIR, dbName);
  try {
    const data = await fs.readFile(filePath, 'utf-8');
  } catch (e: any) {
    await fs.writeFile(filePath, '[]');
  }
}

async function clearDb(dbName: string) {
  await fs.writeFile(path.resolve(DB_DIR, dbName), '[]', 'utf-8');
}

async function loadDb<T>(dbName: string) {
  const fileContent = await fs.readFile(path.resolve(DB_DIR, dbName), {
    encoding: 'utf-8',
    flag: 'rs+',
  });
  const data: T[] = tryParseJson(fileContent, []);

  return data;
}

async function insertOne<T extends RecordType>(dbName: string, payload: T) {
  const dataToInsert = { ...payload };

  const fileContent = await fs.readFile(path.resolve(DB_DIR, dbName), {
    encoding: 'utf-8',
    flag: 'rs+',
  });

  const data: T[] = tryParseJson(fileContent, []);

  dataToInsert.id = dataToInsert.id || randomUUID();
  dataToInsert.createdAt = dayjs(new Date()).format();
  dataToInsert.updatedAt = dayjs(new Date()).format();

  data.push(dataToInsert);

  await fs.writeFile(path.resolve(DB_DIR, dbName), JSON.stringify(data, null, 2), 'utf-8');

  return dataToInsert;
}

async function update<T extends RecordType>(dbName: string, payload: T) {
  const dataToInsert = { ...payload };

  const fileContent = await fs.readFile(path.resolve(DB_DIR, dbName), {
    encoding: 'utf-8',
    flag: 'rs+',
  });
  const data: T[] = tryParseJson(fileContent, []);
  const foundIdx = data.findIndex((i) => i.id === dataToInsert.id);

  dataToInsert.updatedAt = dayjs(new Date()).format();
  data.splice(foundIdx, 1, dataToInsert);

  await fs.writeFile(path.resolve(DB_DIR, dbName), JSON.stringify(data, null, 2), 'utf-8');

  return dataToInsert;
}

export const dbModel = { prepareDb, clearDb, loadDb, insertOne, update };

function readFile(path: string) {}
