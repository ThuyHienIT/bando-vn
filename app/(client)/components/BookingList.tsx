'use client';

import { Col, Empty, Row, Typography } from 'antd';
import Link from 'next/link';
import { useMemo } from 'react';
import styled from 'styled-components';

import { DoubleRightOutlined } from '@ant-design/icons';

import { BookingItem } from './BookingItem';

const ColStyle = styled(Col)``;

const ListStyle = styled(Row)`
  ${ColStyle} {
    margin-bottom: 16px;
  }
`;

const LinkRow = styled.div`
  text-align: right;
`;

interface Props {
  data: BookingItem[];
  heading: string;
  max?: number;
  viewAllLink?: string;
  className?: string;
  onCancel?(item: BookingItem): void;
  onEdit?(item: BookingItem): void;
}

export function BookingList(props: Props) {
  const listData = useMemo(() => {
    if (!props.max || !props.data || props.data.length === 0) return props.data;

    return props.data.slice(0, props.max);
  }, [props.data, props.max]);

  return (
    <div className={props.className}>
      <Typography.Title level={4}>{props.heading}</Typography.Title>
      {listData.length === 0 && (
        <Empty description={`You haven't booked any ${props.heading}`} />
      )}
      {listData.length > 0 && (
        <ListStyle gutter={16}>
          {listData.map((item) => (
            <ColStyle key={item.id} md={6}>
              <BookingItem
                data={item}
                onCancel={props.onCancel}
                onEdit={props.onEdit}
              />
            </ColStyle>
          ))}
        </ListStyle>
      )}

      {props.max && props.data.length > props.max && props.viewAllLink && (
        <LinkRow>
          <Link data-testid="view-all" href={props.viewAllLink}>
            View all <DoubleRightOutlined />
          </Link>
        </LinkRow>
      )}
    </div>
  );
}
