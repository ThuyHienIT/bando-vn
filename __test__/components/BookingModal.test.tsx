import '@testing-library/jest-dom';

import {
  generateBooking,
  generateFacility,
  userInitialState,
} from '__test__/helpers';
import { mockFetch } from '__test__/lib/fetch';
import { BookingModal } from 'app/facility/[id]/components/BookingModal';
import dayjs from 'dayjs';
import { RecoilRoot } from 'recoil';

import { FacilityTypeEnum } from '@enums';
import { dbModel } from '@models/db';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';

const FAC_DB_NAME = 'facilities.json';
const BOOKING_DB_NAME = 'bookings.json';
let BOOKING: BookingItem;

beforeAll(async () => {
  await dbModel.clearDb(FAC_DB_NAME);
  await dbModel.clearDb(BOOKING_DB_NAME);

  const fac1: FacilityItem = generateFacility(FacilityTypeEnum.Room);

  await dbModel.prepareDb(FAC_DB_NAME);
  const insertedFac1 = await dbModel.insertOne<FacilityItem>(FAC_DB_NAME, fac1);

  const booking1 = generateBooking(insertedFac1.id);

  await dbModel.prepareDb(BOOKING_DB_NAME);
  await dbModel.insertOne<BookingItem>(BOOKING_DB_NAME, booking1);
  BOOKING = booking1;
});

beforeEach(() => {
  mockFetch.mockClear();
});

afterAll(async () => {
  await dbModel.clearDb(FAC_DB_NAME);
  await dbModel.clearDb(BOOKING_DB_NAME);
});

describe('Booking Modal List', () => {
  it('smapshot', () => {
    const fac = generateFacility(FacilityTypeEnum.Room);
    fac.id = 'test';
    const { container } = render(
      <RecoilRoot initializeState={userInitialState}>
        <BookingModal
          from={dayjs('2022-02-02')}
          to={dayjs('2022-02-02').add(30, 'minute')}
          opened
          bookingId={BOOKING.id}
          onClose={jest.fn()}
          onDone={jest.fn()}
          onCancelled={jest.fn()}
        />
      </RecoilRoot>
    );

    expect(container).toMatchSnapshot();
  });

  it('verify submit booking', async () => {
    const fac = generateFacility(FacilityTypeEnum.Room);
    fac.id = 'test';
    const from = dayjs();
    const to = from.clone().add(30, 'minute');
    const onClose = jest.fn();

    act(() => {
      render(
        <RecoilRoot initializeState={userInitialState}>
          <BookingModal
            from={from}
            to={to}
            opened
            bookingId={BOOKING.id}
            onClose={onClose}
            onDone={jest.fn()}
            onCancelled={jest.fn()}
          />
        </RecoilRoot>
      );
    });

    const bookingFormEl = await screen.findByTestId('booking-form');
    expect(bookingFormEl).toBeInTheDocument();

    const toInput = bookingFormEl.querySelector('#to') as HTMLInputElement;
    expect(toInput.value).toBe(to.format('HH:mm'));

    const fromInput = bookingFormEl.querySelector('#from') as HTMLInputElement;
    expect(fromInput.value).toBe(from.format('HH:mm'));

    const yesBtn = screen.queryByTestId('booking-modal-yes-btn');
    expect(yesBtn).toBeInTheDocument();
    if (yesBtn) fireEvent.click(yesBtn);

    const noBtn = screen.queryByTestId('booking-modal-cancel-btn');
    expect(noBtn).toBeInTheDocument();
    if (noBtn) fireEvent.click(noBtn);

    await waitFor(() => {
      expect(onClose).toBeCalled();
    });
  });

  it('verify update booking', async () => {
    const fac = generateFacility(FacilityTypeEnum.Room);
    fac.id = 'test';
    const from = dayjs();
    const to = from.clone().add(30, 'minute');
    const container = document.createElement('div');
    const onClose = jest.fn();
    const onDone = jest.fn();
    const onCancelled = jest.fn();
    act(() => {
      render(
        <RecoilRoot initializeState={userInitialState}>
          <BookingModal
            from={from}
            to={to}
            opened
            bookingId={BOOKING.id}
            onClose={onClose}
            onDone={onDone}
            onCancelled={onCancelled}
            isUpdate
            data={fac}
          />
        </RecoilRoot>,
        { container }
      );
    });

    const bookingFormEl = await screen.findByTestId('booking-form');
    expect(bookingFormEl).toBeInTheDocument();

    const toInput = bookingFormEl.querySelector('#to') as HTMLInputElement;
    expect(toInput.value).toBe(to.format('HH:mm'));

    const fromInput = bookingFormEl.querySelector('#from') as HTMLInputElement;
    expect(fromInput.value).toBe(from.format('HH:mm'));

    const yesBtn = screen.queryByTestId('booking-modal-yes-btn');
    expect(yesBtn).toBeInTheDocument();
    if (yesBtn) fireEvent.click(yesBtn);

    await waitFor(() => {
      expect(onDone).toBeCalled();
    });

    const noBtn = screen.queryByTestId('booking-modal-cancel-btn');
    expect(noBtn).toBeInTheDocument();
    if (noBtn) fireEvent.click(noBtn);

    await waitFor(async () => {
      expect(
        await screen.findByText('Cancel your booking')
      ).toBeInTheDocument();
    });
  });
});
