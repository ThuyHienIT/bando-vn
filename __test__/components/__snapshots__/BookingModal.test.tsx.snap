// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking Modal List smapshot 1`] = `<div />`;
